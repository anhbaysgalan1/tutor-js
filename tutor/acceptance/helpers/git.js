const gitP = require('simple-git/promise');
const path = require('path');
const fs = require('fs-extra');
const { imagesDir } = require('./images-dir');
const chalk = require('chalk');

const repoURL = 'https://github.com/openstax/tutor-images.git';

function initialiseRepo(git) {
  console.log(chalk.yellow(`Cloning images repo to ${imagesDir}`));
  return git.clone(repoURL, imagesDir);
}

function getCurrentBranch() {
  const git = gitP(__dirname);
  return new Promise((resolve, reject) => {
    git.branchLocal((err, branch) => {
      if (err) { reject(err); }
      resolve(branch.current);
    });
  });
}

async function gitCommit() {
  const g = git();
  const currentBranchName = await getCurrentBranch();
  const branchInfo = await g.branch();
  if (branchInfo.all.includes(currentBranchName)) {
    await g.checkout(currentBranchName);
  } else {
    await g.checkoutLocalBranch(currentBranchName);
  }
  await g.add(['--all', '.']);
  await g.commit(`updated ${currentBranchName}`);
  await g.push('origin', currentBranchName);
}

async function gitInit() {
  await fs.ensureDir(imagesDir);
  const git = gitP(imagesDir);
  return git.cwd(imagesDir)
    .then(() => git.checkIsRepo())
    .then(isRepo => isRepo ? git : initialiseRepo(git))
    .then(() => git.fetch())
    .then(() => git);
}

function git() {
  process.exit(1)
  return gitP(imagesDir);
}

module.exports = { git, gitCommit, gitInit, getCurrentBranch };
