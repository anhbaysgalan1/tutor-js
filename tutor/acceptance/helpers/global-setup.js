const Server = require('./server');
const puppeteer = require('puppeteer');
const fs = require('fs-extra');
const del = require('del');
const path = require('path');
const { gitInit, gitCommit } = require('./git');
const repoURL = 'https://github.com/openstax/tutor-images.git';

const DIR = require('./working-directory');

async function boot () {
  await gitInit();
  const browser = await puppeteer.launch();
  global.__BROWSER__ = browser;
  fs.ensureDirSync(DIR);
  fs.writeFileSync(path.join(DIR, 'wsEndpoint'), browser.wsEndpoint());

  global.global.__SERVER__ = new Server();
  await global.__SERVER__.boot();
}

module.exports = boot;
