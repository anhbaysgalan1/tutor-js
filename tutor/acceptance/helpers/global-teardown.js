const { gitCommit } = require('./git');
const Server = require('./server');
const DIR = require('./working-directory');
const fs = require('fs-extra');

async function teardown () {
  await global.__BROWSER__.close();
  await global.__SERVER__.halt();
  fs.removeSync(DIR);
  return gitCommit();
}

module.exports = teardown;
