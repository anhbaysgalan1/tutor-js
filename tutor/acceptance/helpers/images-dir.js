const imagesDir = process.env.SCREENSHOTS_DIR;

module.exports = { imagesDir };
