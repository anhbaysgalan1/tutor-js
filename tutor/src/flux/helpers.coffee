_ = require 'underscore'
flux = require 'flux-react'

LOADING = 'loading'
LOADED  = 'loaded'
FAILED  = 'failed'
SAVING  = 'saving'
DELETING = 'deleting'
DELETED = 'deleted'

STATES =
  LOADING : LOADING
  LOADED  : LOADED
  FAILED  : FAILED
  SAVING  : SAVING
  DELETING: DELETING
  DELETED : DELETED

idCounter = 0
CREATE_KEY = -> "_CREATING_#{idCounter++}"

isNew = (id) -> /_CREATING_/.test(id)

CrudConfig = ->
  {
    _asyncStatus: {}
    _local: {}
    _changed: {}
    _errors: {}
    _reload: {}

    # If the specific type needs to do something else to the object:
    # _loaded : (obj, id) ->
    # _saved : (obj, id) ->
    # _reset : ->

    reset: ->
      @_asyncStatus = {}
      @_local = {}
      @_changed = {}
      @_errors = {}
      @_reload = {}

      @_reset?()
      @emitChange()

    dontReload: (id) ->
      @_asyncStatus[id] is LOADED and @_HACK_DO_NOT_RELOAD

    FAILED: (status, msg, id, args...) ->
      @_asyncStatus[id] = FAILED
      @_errors[id] = msg
      @_failed?({status, msg}, id, args...)

      unless status is 0 # indicates network failure
        delete @_local[id]
        @emitChange()

    load: (id) ->
      @_load(id)

    # allow aliasing load without triggering api call.
    _load: (id) ->
      # Add a shortcut for unit testing
      return if @dontReload(id)
      @_reload[id] = false
      @_asyncStatus[id] = LOADING
      @emitChange()

    loaded: (obj, id, args...) ->
      # id = obj.id
      @_asyncStatus[id] = LOADED

      if obj
        # If the specific type needs to do something else to the object:
        @_local[id] = @_loaded?(obj, id, args...) or obj

      @emitChange()

    save: (id, obj) ->
      # Note: id could be isNew()
      @_asyncStatus[id] = SAVING
      @emitChange()

    saved: (result, id, args...) ->
      # id = result.id
      @_asyncStatus[id] = LOADED # TODO: Maybe make this SAVED

      # If the specific type needs to do something else to the object:
      obj = @_saved?(result, id, args...)
      result = obj if obj

      if result
        @_local[id] = result
        @_local[result.id] = result
        delete @_changed[result.id]
      else
        console.warn('API WARN: Server did not return JSON after saving. Patching locally')
        # Merge all the local changes into the new local object
        @_local[id] = _.extend(@_local[id], @_changed[id])
      delete @_changed[id]
      delete @_errors[id]
      # If the specific type needs to do something else to the object:
      @emitChange()

    create: (localId, attributes = {}) ->
      throw new Error('BUG: MUST provide a local id') unless isNew(localId)
      @_local[localId] = {}
      @_changed[localId] = attributes
      @_asyncStatus[localId] = LOADED

    created: (result, localId, args...) ->
      @_local[localId] = result # HACK: So react component can still manipulate the same object
      @_local[result.id] = result
      @_asyncStatus[localId] = LOADED
      @_asyncStatus[result.id] = LOADED
      @emitChange()
      @_created?(result, localId, args...)

    _change: (id, obj) ->
      @_changed[id] ?= {}
      _.extend(@_changed[id], obj)
      @emitChange()

    _save: (id) ->
      @_asyncStatus[id] = SAVING

    delete: (id) ->
      @_asyncStatus[id] = DELETING
      @_delete?(id)

    deleted: (result, id, args...) ->
      @_asyncStatus[id] = DELETED
      delete @_local[id]
      @emitChange()
      @_deleted?(result, id, args...)

    clearChanged: (id) ->
      delete @_changed[id]

    HACK_DO_NOT_RELOAD: (bool) -> @_HACK_DO_NOT_RELOAD = bool

    # Keep this here so other exports method have access to it
    _get: (id) ->
      val = @_local[id]
      return null unless val or @_asyncStatus[id] is SAVING
      # Scores stores an Array unlike most other stores
      if _.isArray(val)
        val
      else
        _.extend({}, val, @_changed[id])

    exports:
      isUnknown: (id) -> not @_asyncStatus[id]
      isLoading: (id) -> @_asyncStatus[id] is LOADING
      isLoaded: (id) -> @_asyncStatus[id] is LOADED
      isDeleting: (id) -> @_asyncStatus[id] is DELETING
      isSaving: (id) -> @_asyncStatus[id] is SAVING
      isFailed: (id) -> @_asyncStatus[id] is FAILED
      getAsyncStatus: (id) -> @_asyncStatus[id]
      get: (id) -> @_get(id)
      isChanged: (id) -> not _.isEmpty(@_changed[id])
      getChanged: (id) -> @_changed[id] or {}
      freshLocalId: -> CREATE_KEY()
      isNew: (id) -> isNew(id)
      reload: (id) -> @_reload[id]
      getError: (id) -> @_errors[id]

  }

# Helper for creating a simple store for actions
makeSimpleStore = (storeConfig) ->

  actionsConfig = _.omit storeConfig, (value, key) ->
    typeof value isnt 'function' or key is 'exports'

  actionsConfig = _.keys(actionsConfig)
  actions = flux.createActions(actionsConfig)
  storeConfig.actions = _.values(actions)
  store = flux.createStore(storeConfig)
  {actions, store}


extendConfig = (newConfig, origConfig) ->
  newConfig.exports ?= {}
  _.defaults(newConfig, origConfig)
  _.defaults(newConfig.exports, origConfig.exports)
  newConfig


makeStandardStore = (name, storeConfig) ->
  extendConfig(storeConfig, new CrudConfig())
  {actions, store} = makeSimpleStore(storeConfig)
  {"#{name}Actions":actions, "#{name}Store":store}


module.exports = {CrudConfig, makeSimpleStore, extendConfig, makeStandardStore, isNew, STATES}
