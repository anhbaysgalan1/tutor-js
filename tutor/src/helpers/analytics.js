import { partial, get } from 'lodash';

import Router from '../helpers/router';
import Courses from '../models/courses-map';

// generate custom event data for routes
const Events = {
  viewTaskStep({ courseId }) {
    // track all work done on a course
    return Analytics.sendEvent('Course', 'Work', { label: courseId });
  },

  viewStudentDashboard({ courseId }) {
    // compare activity between courses
    return Analytics.sendEvent('Student', 'Dashboard', { label: courseId });
  },
};

// a bit shorter helper methods
const isTeacher = courseId => Courses.get(courseId).is_teacher;
const getRole = function(courseId) {
  if (Courses.get(courseId).is_teacher) { return 'teacher'; } else { return 'student'; }
};

const assignmentTypeTranslator = function(assignmentType, { courseId, id }) {
  const type = id === 'new' ? 'create' : 'edit';
  return `/teacher/assignment/${type}/${assignmentType}/${courseId}`;
};

// Translators convert a url like '/foo/bar/123/baz/1' into a simplified one like just '/foo/bar'
const Translators = {
  dashboard({ courseId }) {
    if (isTeacher(courseId)) { return `/teacher/calendar/${courseId}`; } else { return `/student/dashboard/${courseId}`; }
  },
  practiceTopics({ courseId }) { return `/student/practice/${courseId}`; },
  viewPerformanceGuide({ courseId }) {
    return `/${getRole(courseId)}/performance-forecast/${courseId}`;
  },
  calendarByDate({ courseId }) { return `/teacher/calendar/${courseId}`; },
  viewScores({ courseId }) { return `/teacher/student-scores/${courseId}`; },
  courseSettings({ courseId }) { return `/teacher/roster/${courseId}`; },
  editReading:    partial(assignmentTypeTranslator, 'reading'),
  editHomework:   partial(assignmentTypeTranslator, 'homework'),
  editExternal:   partial(assignmentTypeTranslator, 'external'),
  editEvent:      partial(assignmentTypeTranslator, 'event'),
  createReading:  partial(assignmentTypeTranslator, 'reading'),
  createHomework: partial(assignmentTypeTranslator, 'homework'),
  createExternal: partial(assignmentTypeTranslator, 'external'),
  createEvent:    partial(assignmentTypeTranslator, 'event'),
  calendarViewPlanStats({ courseId }) { return `/teacher/metrics/quick/${courseId}`; },
  reviewTask({ courseId }) { return `/teacher/metrics/review/${courseId}`; },
  viewReferenceBook({ courseId }) { return `/reference-view/${courseId}`; },
  viewReferenceBookSection({ courseId, section }) { return `/reference-view/${courseId}/section/${section}`; },
  viewReferenceBookPage({ courseId, cnxId }) { return `/reference-view/${courseId}/page/${cnxId}`; },

  // Task steps are viewed by both teacher and student with no difference in params
  viewTaskStep({ courseId }) {
    const role = Courses.get(courseId).primaryRole.type;
    return `/${role}/task-step/${courseId}`;
  },
};


let GA = undefined;

var Analytics = {

  setTracker(tracker) { return GA = tracker; },

  sendPageView(url) {
    if (GA) { GA('send', 'pageview', url); }
  },

  onNavigation(path) {
    if (!GA) { return; }
    const route = Router.currentMatch(path);

    if (!route) { return this.sendPageView(`/not-found${path}`); }

    const translatedPath = Translators[route.entry.name] ?
          Translators[route.entry.name]( route.params ) : route.pathname;

    // if we're also going to send custom events then we set the page
    if (Events[route.entry.name]) {
      GA('set', 'page', translatedPath);
      Events[route.entry.name]( route.params );
      return this.sendPageView(); // url's not needed since it was set before events
    } else {
      return this.sendPageView(translatedPath);
    }
  },

  sendEvent(category, action, attrs) {
    if (!GA) { return; }
    GA('send', 'event', category, action, attrs.label, attrs.value);
  },
};

export default Analytics;
