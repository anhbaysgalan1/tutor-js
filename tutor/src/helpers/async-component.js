import React from 'react';
import { Button } from 'react-bootstrap';
import { asyncComponent as asyncLoader } from 'react-async-component';
import OXColoredStripe from 'shared/src/components/ox-colored-stripe';
import LoadingScreen from '../components/loading-screen';

const ErrorComponent = ({ error, retry }) => (
  <div className="invalid-page">
    <OXColoredStripe />
    <h1>
      Uh-oh, the page failed to load
    </h1>
    <p>{String(error)}</p>
    <Button bsStyle="primary" onClick={retry}>Retry</Button>
  </div>
);


export function asyncComponent(resolve) {
  return asyncLoader({
    resolve,
    LoadingComponent: LoadingScreen,
    ErrorComponent,
  });
}

export function loadAsync(resolve) {
  // return a function so the router will only evaluate it when it's needed
  // in the future we can insert role dependant logic here
  return () => asyncComponent(resolve);
}
