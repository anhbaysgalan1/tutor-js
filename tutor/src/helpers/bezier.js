import { defer } from 'lodash';

function EASE_IN_OUT(t) {
  if (t < .5) { return 4 * t * t * t; } else { return ((t - 1) * ((2 * t) - 2) * ((2 * t) - 2)) + 1; }
}

function CALC(start, end, elapsed, duration) {
  if (elapsed > duration) { return end; }
  return start + ((end - start) * EASE_IN_OUT(elapsed / duration));
}

const requestAnimationFrame = window.requestAnimationFrame || defer;

export default function bezier({
  range, duration, onStep, onComplete,
}) {
  const startTime = Date.now();

  const step = () => {
    const elapsed = Date.now() - startTime;
    onStep(CALC(range[0], range[1], elapsed, duration));
    if (elapsed < duration) {
      requestAnimationFrame(step);
    } else if (onComplete) {
      onComplete();
    }
  };

  step();
}
