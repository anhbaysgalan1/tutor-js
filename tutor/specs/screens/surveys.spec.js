import { SnapShot, Wrapper } from '../components/helpers/component-testing';
import { bootstrapCoursesList } from '../courses-test-data';
import { StylesManager } from 'survey-react';

import Survey from '../../src/screens/surveys';
import SURVEY from '../../api/research_survey.json';
import EnzymeContext from '../components/helpers/enzyme-context';


describe('Surveys Screen', () => {

  let props, course, surveyRecord;

  beforeEach(() => {
    StylesManager.findSheet = jest.fn();
    props = {
      params: {
        courseId: '1',
        surveyId: '1',
      },
    };
    course = bootstrapCoursesList().get(1);
    course.studentTasks.fetch = jest.fn();

    course.studentTasks.onLoaded({
      data: { tasks: [],
        research_surveys: [SURVEY],
      },
    });
    surveyRecord = course.studentTasks.researchSurveys.get('1');
  });


  it('renders and matches snapshot', () => {
    const component = SnapShot.create(<Wrapper _wrapped_component={Survey} {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('submits survey when answered', () => {
    const survey = mount(<Survey {...props} />, EnzymeContext.build());
    survey.find('input[type="radio"]').first().simulate('change',
      { target: { value: 'bananas' } }
    );
    surveyRecord.save = jest.fn();
    survey.find('.sv_complete_btn').simulate('click', {});
    expect(surveyRecord.save).toHaveBeenCalled();
    expect(surveyRecord.response).toEqual({ favs: 'bananas' });
    survey.unmount();
  });

});
