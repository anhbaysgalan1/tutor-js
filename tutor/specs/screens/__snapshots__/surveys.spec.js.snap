// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Surveys Screen renders and matches snapshot 1`] = `
<div
  className="course-page research-surveys"
  data-appearance="testing"
  data-book-title="Testing"
  data-is-preview={false}
  data-term="Spring 2017"
  data-title="Local Test Course One"
>
  <header>
    <div
      className="title-wrapper"
    >
      <div
        className="title"
      >
        Local Test Course One
      </div>
      <div
        className="subtitle"
      >
        Spring 2017
      </div>
    </div>
  </header>
  <div
    className="body-wrapper"
  >
    <div
      className="body"
    >
      <h1>
        Survey “
        What are your favorites?
        ”
      </h1>
      <div
        className="sv_main sv_default_css"
      >
        <div
          className="sv_custom_header"
        />
        <div
          className="sv_container"
        >
          <div
            className="sv_header"
          >
            <h3 />
          </div>
          <div
            className="sv_body"
            id="sp_100"
          >
            <div
              className="sv_p_root"
            >
              <div
                className="sv_row"
              >
                <div
                  className="sv_q sv_qstn"
                  id="sq_100"
                  style={
                    Object {
                      "width": "100%",
                    }
                  }
                >
                  <div
                    className={null}
                  >
                    <h5
                      className="sv_q_title"
                    >
                      <span
                        style={null}
                      >
                        1. Which fruit is best?
                      </span>
                    </h5>
                  </div>
                  <div
                    className={null}
                  >
                    <fieldset
                      className="sv_qcbc"
                    >
                      <div
                        className="sv_q_radiogroup"
                        style={
                          Object {
                            "display": "inline-block",
                            "width": "100%",
                          }
                        }
                      >
                        <label
                          className="sv_q_radiogroup_label"
                        >
                          <input
                            aria-label="1. Which fruit is best?"
                            checked={false}
                            className="sv_q_radiogroup_control_item"
                            disabled={false}
                            id="sq_100i"
                            name="favs_sq_100"
                            onChange={[Function]}
                            type="radio"
                            value="bananas"
                          />
                          <span
                            className="circle"
                          />
                          <span
                            className="check"
                          />
                          <span
                            className=""
                          >
                            <span
                              style={
                                Object {
                                  "display": "inline",
                                  "marginLeft": "3px",
                                  "position": "static",
                                }
                              }
                            >
                              Bananas
                            </span>
                          </span>
                        </label>
                      </div>
                      <div
                        className="sv_q_radiogroup"
                        style={
                          Object {
                            "display": "inline-block",
                            "width": "100%",
                          }
                        }
                      >
                        <label
                          className="sv_q_radiogroup_label"
                        >
                          <input
                            aria-label="1. Which fruit is best?"
                            checked={false}
                            className="sv_q_radiogroup_control_item"
                            disabled={false}
                            id={null}
                            name="favs_sq_100"
                            onChange={[Function]}
                            type="radio"
                            value="oranges"
                          />
                          <span
                            className="circle"
                          />
                          <span
                            className="check"
                          />
                          <span
                            className=""
                          >
                            <span
                              style={
                                Object {
                                  "display": "inline",
                                  "marginLeft": "3px",
                                  "position": "static",
                                }
                              }
                            >
                              Oranges
                            </span>
                          </span>
                        </label>
                      </div>
                      <div
                        className="sv_q_radiogroup"
                        style={
                          Object {
                            "display": "inline-block",
                            "width": "100%",
                          }
                        }
                      >
                        <label
                          className="sv_q_radiogroup_label"
                        >
                          <input
                            aria-label="1. Which fruit is best?"
                            checked={false}
                            className="sv_q_radiogroup_control_item"
                            disabled={false}
                            id={null}
                            name="favs_sq_100"
                            onChange={[Function]}
                            type="radio"
                            value="apples"
                          />
                          <span
                            className="circle"
                          />
                          <span
                            className="check"
                          />
                          <span
                            className=""
                          >
                            <span
                              style={
                                Object {
                                  "display": "inline",
                                  "marginLeft": "3px",
                                  "position": "static",
                                }
                              }
                            >
                              Apples
                            </span>
                          </span>
                        </label>
                      </div>
                      <legend
                        style={
                          Object {
                            "display": "none",
                          }
                        }
                      >
                        1. Which fruit is best?
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="sv_nav"
            >
              <input
                className=" sv_complete_btn"
                onClick={[Function]}
                style={
                  Object {
                    "marginRight": "5px",
                  }
                }
                type="button"
                value="Complete"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
